### PR TITLE
Use the _get_comm_info method to reconstruct all widget models

### LINKED ISSUE #133: Persistence issues with widgets holding other widgets
#### Re-instantiating all valid widget models
Let' s say that we have a container widget with id `id1` holding two widgets of id `id2` and `id3`. (For example, `HBox([Button(), Button()]))`. We save the notebook and the states of the three widgets are saved to the local storage.

``` Python
id1
 |
 +--- id2
 |
 +--- id3
```

Then we change the content initial container wigdet, so that we are in the situation

``` Python
id1
 |
 +--- id4
 |
 +--- id5
```

then, without saving the notebook, we reload the page.

Since `id1` is a valid comm, we request it' s state, but when de-serializing its state, we fail at instantiating the widget models `id4` , `id5` because they were not recreated. (more precisely, get_model will return undefined.) Errors are displayed in the console.

This shows that we should re-instantiate all the valid widget models regardless of whether they were in the local storage. 

#### External state request

The issue is that we need to retrieve the `_model_name` and the `_model_module`, which are part of the state to instantiate the model and make a state request.

Hence, the state request should probably be external to the widget, and sent via the bare comm object.

### LINKED ISSUE #155: reconnect to widget
currently, when starting a long-running process i’d like to close my laptop lid, open it later, and see my progressbar jump to its current position.

the notebook itself handles reconnection well, but the widgets simply vanish.

### PR DESCRIPTION
**Pull request by jdfreder**
Seems to fix a lot of bugs with persistence in general.

@SylvainCorlay woohoo!

closes #133
closes #155

### CODE REVIEW

**Comment by SylvainCorlay** [comment]
@jdfreder woohoo. I will look at it tonight.

File: ipywidgets/static/widgets/js/manager.js [hunk 1]
```diff
@@ -376,17 +382,69 @@ define([
...
+            });
+            
+            // Send a state request message out for each widget comm and wait
+            // for the responses.
+            return Promise.all(comm_promises).then(function(comms) {
+                return Promise.all(comms.map(function(comm) {
+                    var update_promise = new Promise(function(resolve, reject) {
+                        comm.on_msg(function (msg) {
+                            
+                            // A suspected response was recieved, check to see if
```
Comment (SylvainCorlay) [inline ipywidgets/static/widgets/js/manager.js]: typo

File: ipywidgets/static/widgets/js/manager.js [hunk 2]
```diff
@@ -317,6 +317,12 @@ define([
         // Recreate all the widget models for the given notebook state.
         var all_models = that._get_comm_info().then(function(live_comms) {
             return Promise.all(_.map(Object.keys(state), function (model_id) {
+                
+                // If the model has already been created, return it.
+                if (that._models[model_id]) {
+                    return that._models[model_id];
+                }
+
```
Comment (SylvainCorlay) [inline ipywidgets/static/widgets/js/manager.js]: In which situation could the model already be created?

**Comment by SylvainCorlay** [comment]
@jdfreder the `comm_info_[request/reply]` shell message allows you to only request comms corresponding to a given target name. (You can see this in #75.)  
It would make sense to make the workaround that we have here do the same thing so that it will be easier to replace when the notebook 4.1 is released like here: https://github.com/SylvainCorlay/ipython_widgets/commit/78984210255f71f1e01c01f331fb978a75feebb6

File: ipywidgets/static/widgets/js/manager.js (hunk 2 above)
Comment (jdfreder) [inline ipywidgets/static/widgets/js/manager.js]: In line 425 of the block below.  Before I call this function I create models based on `_get_comm_info` return info.

Comment (SylvainCorlay) [inline ipywidgets/static/widgets/js/manager.js]: oh, ok

**Comment by jdfreder** [comment]
@SylvainCorlay comments addressed

**Comment by SylvainCorlay** [comment]
:+1: :tada: :balloon: 
I am all good with this. I can merge later this afternoon, just to let the time to other people to comment. Any input @jasongrout?

**Comment by SylvainCorlay** [comment]
hop?

**Pull request merged**